### ISSUE TITLE
KeyError in nonlinsolve

### ISSUE THREAD

**Issue by ytyt-yt** [sympy/sympy#21022]
I got a weird error with `nonlinsolve`:
```python
import sympy as sy


eqs = [
    'k-16',
    'p-8',
    'y*y+z*z-x*x',
    'd - x + p',
    'd*d+k*k-y*y',
    # 'p*p+k*k-z*z',
    'z*z-p*p-k*k',
    'abc-efg',
]

eqs = [sy.sympify(x) for x in eqs]

# syb = ['x', 'y', 'z', 'k', 'p', 'd', 'abc', 'efg']
syb = list(set.union(*[x.free_symbols for x in eqs]))
res = sy.solvers.nonlinsolve(eqs, syb)
print(res)
```

it will raise:
```
/xxxxxxx/python3.9/site-packages/sympy/core/operations.py:64: SymPyDeprecationWarning: 

Add/Mul with non-Expr args has been deprecated since SymPy 1.7. Use
Expr args instead. See https://github.com/sympy/sympy/issues/19445 for
more info.

  SymPyDeprecationWarning(
Traceback (most recent call last):
  File "/tmp/test.py", line 19, in <module>
    res = sy.solvers.nonlinsolve(eqs, syb)
  File "/xxxxxxx/python3.9/site-packages/sympy/solvers/solveset.py", line 3529, in nonlinsolve
    result = substitution(
  File "/xxxxxxx/python3.9/site-packages/sympy/solvers/solveset.py", line 3194, in substitution
    new_result_real, solve_call1, cnd_call1 = _solve_using_known_values(
  File "/xxxxxxx/python3.9/site-packages/sympy/solvers/solveset.py", line 3172, in _solve_using_known_values
    imgst = soln_imageset[sol]
KeyError: ConditionSet(y, Eq(-z**2/16 + sqrt(y**2 - 256) - 12, 0), FiniteSet(-sqrt((z**2 - 16*z + 320)*(z**2 + 16*z + 320))/16, sqrt((z**2 - 16*z + 320)*(z**2 + 16*z + 320))/16))

```

And I found 3 ways to make the results right:
1. remove the equation `'abc-efg'`
2. manually write down symbols (`syb = ['x', 'y', 'z', 'k', 'p', 'd', 'abc', 'efg']`)
3. replace `z*z-p*p-k*k` with `p*p+k*k-z*z`. (really weird).

All of these 3 modifications should be equivalent to the original code. Is there anything wrong?

**Comment by oscarbenjamin** [comment]
This is a bug in solveset. Somehow it ends up thinking that the conditionset shown is an expression and trying to substitute it for a symbol.

**Comment by ytyt-yt** [comment]
> This is a bug in solveset. Somehow it ends up thinking that the conditionset shown is an expression and trying to substitute it for a symbol.

Thank you for your reply! any plan or idea to solve it?

**Comment by oscarbenjamin** [comment]
Someone would need to go through in a debugger to find where the set gets mixed up with an expression.

**Comment by eagleoflqj** [comment]
This should be reopened.
Solve it manually,
```
k=16
p=8
z**2=320 => z=\pm 8*sqrt(5)
x**2-y**2=320
x=d+8 => x**2=d**2+16*d+64
y**2=d**2+256
320=16*d-192 => d=32 => x=40, y=\pm 16*sqrt(5)
```
There are 4 solutions, but current test case's 4 solutions are:
```py
>>> for arg in ans.args:
...     print([t.simplify() for t in arg])

[efg, 32, efg, 16, 8, 40, 16*sqrt(5), -8*sqrt(5)]
[efg, 32, efg, 16, 8, 40, 16*sqrt(5), 8*sqrt(5)]
[efg, -32, efg, 16, 8, -24, 16*sqrt(5), -8*sqrt(5)]
[efg, -32, efg, 16, 8, -24, 16*sqrt(5), 8*sqrt(5)]
```
The last 2 are wrong, and 2 solutions are missing.

**Comment by oscarbenjamin** [comment]
With #23011 the solutions would be given as:
```
In [4]: for s in nonlinsolve(eqs, syb):
   ...:     pprint(s)
(efg, 8, 32, -8⋅√5, efg, -16⋅√5, 40, 16)
(efg, 8, 32, -8⋅√5, efg, 16⋅√5, 40, 16)
(efg, 8, 32, 8⋅√5, efg, -16⋅√5, 40, 16)
(efg, 8, 32, 8⋅√5, efg, 16⋅√5, 40, 16)

In [5]: syb
Out[5]: [abc, p, d, z, efg, y, x, k]
```
I guess there is a bug in the `substitution` solver but the poly solver can give the correct solution here. Ideally the bug in `substitution` would be fixed as well as making use of the poly solver.

**Comment by smichr** [comment]
A minimal failing example is found as follows:
```python
>>> eqs = [sy.sympify(x) for x in eqs]
>>> for i in subsets(eqs):
...  if not i:continue
...  free = list(set.union(*[_.free_symbols for _ in i]))
...  for j in subsets(free, len(i)):
...    s, ss = solve(i, j, set=1)
...    tt = nonlinsolve(i, j)
...    assert ss == tt
...
Traceback (most recent call last):
  File "<stdin>", line 7, in <module>
AssertionError
>>> i
(k - 16, -x**2 + y**2 + z**2)
>>> j
(x, k)
>>> ss
{(16, -sqrt(y**2 + z**2)), (16, sqrt(y**2 + z**2))}
>>> tt
{(-sqrt(y**2 + z**2), 16), (sqrt(y**2 + z**2), 16)}
```

The order of the results is reversed from nonlinsolve.

**Comment by eagleoflqj** [comment]
Should be fixed by #23011

**Comment by oscarbenjamin** [comment]
Yes and a test is already added:
https://github.com/sympy/sympy/blob/c8f2320f0a593c57509dec24b64d0c3da2164a11/sympy/solvers/tests/test_solveset.py#L2161
